### DISCUSSION TITLE
Is it possible for a new zfs dataset to be created per jail at jail creation time?

### DISCUSSION THREAD

**Discussion by Kailee71** [Jip-Hop/jailmaker#68] [Ideas]
HI all,

at the moment jails "just" live in subdirectories of the jailmaker dataset/zvol. Would it be possible to create a seperate dataset (or zvol for that matter) per jail at creation time? This would enable very straightforward and reliable backup strategies though the snapshot and replication tasks already included in Scale. Would love to see this included.

**Comment by Jip-Hop** [comment]
Hi. You can already snapshot the jailmaker dataset. Is that not enough? I personally keep all the data separated from the jailmaker rootfs directories. I bind mount these datasets containing data, which may require different snapshot policies, into the jail using bind mounts.

Jailmaker doesn't currently create new datasets for each rootfs, but you could create them manually beforehand. First create the jailmaker dataset as instructed in the readme. Then create a jails dataset inside it and inside a myjail dataset and inside it a rootfs dataset. Then you'd have to use jailmaker to create a new jail named e.g. myjail_tmp and upon completion don't start it. Then copy all files from the myjail_tmp jails dir into the myjail dataset. This is obviously a very cumbersome process...

Alternatively I think you can make a symlink for the rootfs directory. Just create a jail as usual. Then create a new dataset somewhere. Copy all the contents of the rootfs dir inside this new dataset. Remove the original rootfs dir and make a symlink from rootfs to the new dataset. 

This is from the top of my head, I did not confirm above suggestions work.

I think it's a nice idea though and if anyone is interested in building this feature then I'd be open to merging it.

**Reply by Kailee71** [reply]
I personally would really love to be able to split individual jails off into own datasets. This would enable finer grained snapshots per jail ("let me quickly do an upgrade on jail XYZ but be sure of a quick return path") which can be super useful if it's just a playground jail...

**Reply by Jip-Hop** [reply]
For a quick return path you can already snapshot the entire jailmaker dataset and copy the rootfs folder from your playground jail from the snapshot folder to restore. If you don't want to copy but instead rollback a snapshot I suggest you try the previous symlink suggestion. Until someone contributes this feature to jailmaker I think those are your best options :)

**Comment by Jip-Hop** [comment]
I'm locking this discussion as I've opened dedicated issues to track feature requests. Please see [Create ZFS dataset for each jail](https://github.com/Jip-Hop/jailmaker/issues/68).

**Reply by Jip-Hop** [reply]
@Kailee71 this is now included in [v1.1.4](https://github.com/Jip-Hop/jailmaker/releases/tag/v1.1.4).
